File: src/flutter-app/TaskTracker/lib/service/team.dart
/*
 * Copyright (c) 2020 by Botorabi. All rights reserved.
 * https://github.com/botorabi/TaskTracker
 *
 * License: MIT License (MIT), read the LICENSE text in
 *          main directory for more details.
 */

import 'dart:convert';

import 'package:TaskTracker/common/utf8.utils.dart';

class Team {

  int id;
  String name = '';
  String description = '';
  bool active = false;
  List<int> userIDs = [];
  List<int> teamLeaderIDs = [];
  List<String> teamLeaderNames = [];

  Team();

  factory Team.fromMap(final Map<String, dynamic> fields) {
    Team team = Team();
    team.id = fields['id'];
    team.name = Utf8Utils.fromUtf8(fields['name']);
    team.description = Utf8Utils.fromUtf8(fields['description']);
    if (fields.containsKey('active')) {
      team.active = fields['active'];
    }
    if (fields.containsKey('userIDs')) {
      team.userIDs = List.from(fields['userIDs']);
    }
    if (fields.containsKey('teamLeaderIDs')) {
      team.teamLeaderIDs = List.from(fields['teamLeaderIDs']);
    }
    if (fields.containsKey('teamLeaderNames')) {
      team.teamLeaderNames = List.from(fields['teamLeaderNames']);
    }
    return team;
  }

  factory Team.fromJsonString(final String jsonString) {
    Map<String, dynamic> fields = jsonDecode(jsonString);
    return Team.fromMap(fields);
  }

  static List<Team> listFromJsonString(final String jsonString) {
    List<Team> teamList = List<Team>();
    dynamic teams = jsonDecode(jsonString);
    teams.forEach((element) {
      teamList.add(Team.fromMap(element));
    });
    return teamList;
  }

  Map toJson() {
    return {
      'id' : id,
      'name' : name,
      'description' : description,
      'active' : active,
      'userIDs' : userIDs?.toList(),
      'teamLeaderIDs' : teamLeaderIDs?.toList(),
    };
  }
}

File: src/main/java/net/vrfun/tasktracker/user/Team.java
/*
 * Copyright (c) 2020-2021 by Botorabi. All rights reserved.
 * https://github.com/botorabi/TaskTracker
 *
 * License: MIT License (MIT), read the LICENSE text in
 *          main directory for more details.
 */
package net.vrfun.tasktracker.user;

import net.vrfun.tasktracker.common.BaseEntity;
import org.springframework.lang.*;

import javax.persistence.*;
import java.io.Serializable;
import java.util.Collection;

@Entity
public class Team extends BaseEntity implements Serializable {

    private static final long serialVersionUID = 1L;

    @Id
    @GeneratedValue(strategy = GenerationType.AUTO)
    private Long id;

    @Column(nullable=false)
    private String name;

    @Column(length = 1024)
    private String description;

    private boolean active;

    @ManyToMany(targetEntity = User.class, cascade = {CascadeType.DETACH, CascadeType.REFRESH, CascadeType.MERGE})
    private Collection<User> users;

    @ManyToMany(targetEntity = User.class, cascade = {CascadeType.DETACH, CascadeType.REFRESH, CascadeType.MERGE})
    private Collection<User> teamLeaders;

    public Team() {}

    public Team(@NonNull final String name,
                @Nullable final String description) {

        this.name = name;
        this.description = description;
        this.active = true;
    }

    public Long getId() {
        return id;
    }

    public void setId(Long id) {
        this.id = id;
    }

    @NonNull
    public String getName() {
        return name;
    }

    public void setName(@NonNull String name) {
        this.name = name;
    }

    @Nullable
    public String getDescription() {
        return description;
    }

    public void setDescription(@Nullable String description) {
        this.description = description;
    }

    public boolean getActive() {
        return active;
    }

    public void setActive(boolean active) {
        this.active = active;
    }

    @Nullable
    public Collection<User> getUsers() {
        return users;
    }

    public void setUsers(@Nullable Collection<User> users) {
        this.users = users;
    }

    @Nullable
    public Collection<User> getTeamLeaders() {
        return teamLeaders;
    }

    public void setTeamLeaders(@Nullable Collection<User> teamLeaders) {
        this.teamLeaders = teamLeaders;
    }
}
